### PR TITLE
Alignment configuration free end gaps

### PR DESCRIPTION
**Pull request by smehringer**
Resolves seqan/product_backlog/issues/96

The alignment_configurator adaption is probably not the smartest move but it is part of the old alignment and will be deleted anyway.

### CODE REVIEW

File: include/seqan3/alignment/pairwise/detail/policy_alignment_matrix.hpp [hunk 1]
```diff
@@ -100,23 +100,20 @@ class policy_alignment_matrix
...
+            last_column_is_free = method_global_config.free_end_gaps_sequence2_trailing.get();
             // band starts in first column without free gaps or band starts in first row without free gaps.
             invalid_band |= (upper_diagonal < 0 && !first_column_is_free) || (lower_diagonal > 0 && !first_row_is_free);
             error_cause += " The band starts in a region without free gaps.";
         }
-        else if constexpr (traits_t::is_global)
-        {
-            invalid_band |= (upper_diagonal < 0 || lower_diagonal > 0);
-            error_cause += " The first cell of the matrix is not enclosed by the band.";
-        }
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/detail/policy_alignment_matrix.hpp:119]: ?

File: include/seqan3/alignment/pairwise/detail/policy_alignment_matrix.hpp [hunk 2]
```diff
@@ -100,23 +100,20 @@ class policy_alignment_matrix
 
         bool invalid_band = upper_diagonal < lower_diagonal;
         std::string error_cause = (invalid_band) ? " The upper diagonal is smaller than the lower diagonal." : "";
-        if constexpr (traits_t::with_free_end_gaps)
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/detail/policy_alignment_matrix.hpp:103]: ?

**Comment by marehr** [comment]
Is every single commit "buildable"? If so could we split this PR into multiple ones, this would make it easier on the reviewers (I would review all of them, because I know somehow now what this big PR is about).

**Comment by smehringer** [comment]
> 1st_place_medal For this very Ren[n]e like PR safety_vest construction_worker_man

I did my best.

> This is just the first round, I have some problem to find the important changes. Even some https://www.youtube.com/watch?v=DU2bgWOln64 couldn't help me xD
> I'll go commit-by-commit again, tomorrow.
> 
> Is this a follow-up PR of other changes that I'm maybe unaware of?

Yes sorry, it is the second part of seqan/product_backlog/issues/96, following PR https://github.com/seqan/seqan3/pull/1938

> Is every single commit "buildable"? 

Yes

> If so could we split this PR into multiple ones, this would make it easier on the reviewers (I would review all of them, because I now know somehow what this big PR is about).

I know this will sound childish: Rene never has to :(
And actually the commits do belong together quite strongly. In the first part in PR https://github.com/seqan/seqan3/pull/1938
I just introduced the `method_global` configuration without using it and without initializing its free end gap members.
In this part I do (by commit):

- Add the free and gap initialisation wherever it is needed in a `method_global` configuration
- Adapt the "new alignment code" to actually use `method_global` instead of `aligned_ends`
- Adapt the "old alignment code" to actually use `method_global` instead of `aligned_ends`
- Delete all now unnecessary usages of the `aligned_ends` config
- Documentation
- Moving `aligned_ends` to `detail::aligned_ends` because it is still needed in the "old alignment code" (this probably fits better after adapt the old alignment code.. I can change that)

So I already tried to split this in commits although it is basically the single task of "replace aligned_ends with method_global".
I thought in single PRs the changes won't have much meaning but I can make single PR's if you like.

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 3]
```diff
@@ -319,14 +320,16 @@ struct alignment_configurator
...
             {
+                auto method_global_cfg = get<seqan3::align_cfg::method_global>(config_with_result_type);
                 // Only use edit distance if ...
                 if (gaps.get_gap_open_score() == 0 &&  // gap open score is not set,
-                    !(align_ends_cfg[2] || align_ends_cfg[3]) && // none of the free end gaps are set for second seq,
-                    align_ends_cfg[0] == align_ends_cfg[1]) // free ends for leading and trailing gaps are equal in first seq.
+                    !(method_global_cfg.free_end_gaps_sequence2_leading.get() ||
+                      method_global_cfg.free_end_gaps_sequence2_trailing.get()) &&
+                    (method_global_cfg.free_end_gaps_sequence1_leading.get() ==
+                     method_global_cfg.free_end_gaps_sequence1_trailing.get()))
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/alignment/pairwise/alignment_configurator.hpp:332]: Oh you are right. Um, I just assumed to also use strong types but I have no strong feelings. Shall I change them to boolean?

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alignment/pairwise/alignment_configurator.hpp:332]: Yes, this should be changed.

**Comment by smehringer** [comment]
superseeded by separate PRs

**Pull request closed without merging**